Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,1150 @@
++        if(players != null) {
++            for (EntityPlayer player : players) {
++                if(player.isSpectator() || !player.affectsSpawning) continue;
++                //Paper end
++                ++player.mobCounts[index];
++            }
+         }
+     }
+ 
+@@ -936,29 +940,31 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+     protected void setViewDistance(int i) {
+         int j = MathHelper.clamp(i + 1, 3, 33);
+ 
+-        if (j != this.viewDistance) {
+-            int k = this.viewDistance;
+-
+-            this.viewDistance = j;
+-            this.chunkDistanceManager.a(this.viewDistance);
+-            ObjectIterator objectiterator = this.updatingChunks.values().iterator();
++        //Paper start ticket level pp-view-dist
++        if(this.world.players != null && !this.world.players.isEmpty()) {
++            for (EntityPlayer player : this.world.players) {
++                player.setViewDistance(j);
++                this.playerTicketAreaMap.updatePlayer(player);
++            }
++        }
++        //Paper end
++    }
+ 
+-            while (objectiterator.hasNext()) {
+-                PlayerChunk playerchunk = (PlayerChunk) objectiterator.next();
+-                ChunkCoordIntPair chunkcoordintpair = playerchunk.i();
++    //Paper start ticket level pp-view-dist
++    public void updateTrackViewDistance(EntityPlayer player, int i) {
++        for (int x = player.chunkX - i; x <= player.chunkX + i; ++x) {
++            for (int z = player.chunkZ - i; z <= player.chunkZ + i; ++z) {
++                ChunkCoordIntPair chunkcoordintpair = new ChunkCoordIntPair(x, z);
+                 Packet<?>[] apacket = new Packet[2];
++                int i1 = b(chunkcoordintpair, player, true);
++                boolean flag = i1 <= player.lastTrackViewDistance;
++                boolean flag1 = i1 <= i;
+ 
+-                this.a(chunkcoordintpair, false).forEach((entityplayer) -> {
+-                    int l = b(chunkcoordintpair, entityplayer, true);
+-                    boolean flag = l <= k;
+-                    boolean flag1 = l <= this.viewDistance;
+-
+-                    this.sendChunk(entityplayer, chunkcoordintpair, apacket, flag, flag1);
+-                });
++                this.sendChunk(player, chunkcoordintpair, apacket, flag, flag1);
+             }
+         }
+-
+     }
++    //Paper end
+ 
+     protected void sendChunk(EntityPlayer entityplayer, ChunkCoordIntPair chunkcoordintpair, Packet<?>[] apacket, boolean flag, boolean flag1) {
+         if (entityplayer.world == this.world) {
+@@ -966,10 +972,13 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+                 PlayerChunk playerchunk = this.getVisibleChunk(chunkcoordintpair.pair());
+ 
+                 if (playerchunk != null) {
+-                    Chunk chunk = playerchunk.getChunk();
++                    Chunk chunk = playerchunk.getTickingChunk(); //for clarification
```

Comment:
this diff isn't technically needed, the else should cover explanation

Reply:
solved in rewrite